File: CS/RealTimeChartUpdates/Form1.cs
using DevExpress.Utils;
using DevExpress.XtraCharts;
using System;
using System.Collections.Generic;
using System.Collections.ObjectModel;
using System.ComponentModel;
using System.Threading;
using System.Windows.Forms;

namespace RealTimeChartUpdates {
    public partial class Form1 : Form {
        const int ViewportPointCount = 5000;
        int counter = 0;
        Thread dataAcquisitionThread;
        bool isEnabled = true;
        int lastRawDataIndex = 0;
        List<DataPoint> rawData = new List<DataPoint>();
        ObservableCollection<DataPoint> viewportData = new ObservableCollection<DataPoint>();
        public Form1() { InitializeComponent(); }
        void Form1_Load(object sender, EventArgs e) {
            dataAcquisitionThread = new Thread(new ThreadStart(AcquireData));
            dataAcquisitionThread.Start();
            chartControl1.Titles.Add(new ChartTitle { Text = "Real-Time Charting" });

            Series series = new Series();
            series.ChangeView(ViewType.Line);
            series.DataSource = viewportData;
            series.DataSourceSorted = true;
            series.ArgumentDataMember = "Argument";
            series.ValueDataMembers.AddRange("Value");
            chartControl1.Series.Add(series);

            LineSeriesView seriesView = (LineSeriesView)series.View;
            seriesView.LastPoint.LabelDisplayMode = SidePointDisplayMode.SeriesPoint;
            seriesView.LastPoint.Label.TextPattern = "{V:f2}";

            XYDiagram diagram = (XYDiagram)chartControl1.Diagram;
            diagram.AxisX.DateTimeScaleOptions.ScaleMode = ScaleMode.Continuous;
            diagram.AxisX.Label.ResolveOverlappingOptions.AllowRotate = false;
            diagram.AxisX.Label.ResolveOverlappingOptions.AllowStagger = false;
            diagram.AxisX.VisualRange.EndSideMargin = 200;
            diagram.DependentAxesYRange = DefaultBoolean.True;
            diagram.AxisY.WholeRange.AlwaysShowZeroLevel = false;

            System.Windows.Forms.Timer timer = new System.Windows.Forms.Timer();
            timer.Interval = 100;
            timer.Start();
            timer.Tick += Timer_Tick;
        }
        void AcquireData() {
            while (isEnabled) {
                Thread.Sleep(15);
                lock (rawData) {
                    for (int i = 0; i < 50; i++)
                        rawData.Add(new DataPoint(DateTime.Now, GenerateValue(counter++)));
                }
            }
        }
        double GenerateValue(double x) { return Math.Sin(x / 1000.0) * 3 * x + x / 2 + 5; }
        void Timer_Tick(object sender, EventArgs e) {
            lock (rawData) {
                for (int i = Math.Max(lastRawDataIndex, rawData.Count - ViewportPointCount); i < rawData.Count; i++)
                    viewportData.Add(rawData[i]);
                lastRawDataIndex = rawData.Count;
                while (viewportData.Count > ViewportPointCount)
                    viewportData.RemoveAt(0);
            }
        }
        protected override void OnClosing(CancelEventArgs e) {
            isEnabled = false;
            base.OnClosing(e);
        }
        public class DataPoint {
            public DataPoint(DateTime argument, double value) {
                Argument = argument;
                Value = value;
            }
            public DateTime Argument { get; set; }
            public double Value { get; set; }
        }
    }
}

File: VB/RealTimeChartUpdates/Form1.vb
Imports DevExpress.Utils
Imports DevExpress.XtraCharts
Imports System
Imports System.Collections.Generic
Imports System.Collections.ObjectModel
Imports System.ComponentModel
Imports System.Threading
Imports System.Windows.Forms

Namespace RealTimeChartUpdates

    Public Partial Class Form1
        Inherits Form

        Const ViewportPointCount As Integer = 5000

        Private counter As Integer = 0

        Private dataAcquisitionThread As Thread

        Private isEnabled As Boolean = True

        Private lastRawDataIndex As Integer = 0

        Private rawData As List(Of DataPoint) = New List(Of DataPoint)()

        Private viewportData As ObservableCollection(Of DataPoint) = New ObservableCollection(Of DataPoint)()

        Public Sub New()
            InitializeComponent()
        End Sub

        Private Sub Form1_Load(ByVal sender As Object, ByVal e As EventArgs)
            dataAcquisitionThread = New Thread(New ThreadStart(AddressOf AcquireData))
            dataAcquisitionThread.Start()
            chartControl1.Titles.Add(New ChartTitle With {.Text = "Real-Time Charting"})
            Dim series As Series = New Series()
            series.ChangeView(ViewType.Line)
            series.DataSource = viewportData
            series.DataSourceSorted = True
            series.ArgumentDataMember = "Argument"
            series.ValueDataMembers.AddRange("Value")
            chartControl1.Series.Add(series)
            Dim seriesView As LineSeriesView = CType(series.View, LineSeriesView)
            seriesView.LastPoint.LabelDisplayMode = SidePointDisplayMode.SeriesPoint
            seriesView.LastPoint.Label.TextPattern = "{V:f2}"
            Dim diagram As XYDiagram = CType(chartControl1.Diagram, XYDiagram)
            diagram.AxisX.DateTimeScaleOptions.ScaleMode = ScaleMode.Continuous
            diagram.AxisX.Label.ResolveOverlappingOptions.AllowRotate = False
            diagram.AxisX.Label.ResolveOverlappingOptions.AllowStagger = False
            diagram.AxisX.VisualRange.EndSideMargin = 200
            diagram.DependentAxesYRange = DefaultBoolean.True
            diagram.AxisY.WholeRange.AlwaysShowZeroLevel = False
            Dim timer As System.Windows.Forms.Timer = New System.Windows.Forms.Timer()
            timer.Interval = 100
            timer.Start()
            AddHandler timer.Tick, AddressOf Timer_Tick
        End Sub

        Private Sub AcquireData()
            While isEnabled
                Thread.Sleep(15)
                SyncLock rawData
                    For i As Integer = 0 To 50 - 1
                        rawData.Add(New DataPoint(Date.Now, GenerateValue(Math.Min(Interlocked.Increment(counter), counter - 1))))
                    Next

                End SyncLock
            End While
        End Sub

        Private Function GenerateValue(ByVal x As Double) As Double
            Return Math.Sin(x / 1000.0) * 3 * x + x / 2 + 5
        End Function

        Private Sub Timer_Tick(ByVal sender As Object, ByVal e As EventArgs)
            SyncLock rawData
                For i As Integer = Math.Max(lastRawDataIndex, rawData.Count - ViewportPointCount) To rawData.Count - 1
                    viewportData.Add(rawData(i))
                Next

                lastRawDataIndex = rawData.Count
                While viewportData.Count > ViewportPointCount
                    viewportData.RemoveAt(0)
                End While

            End SyncLock
        End Sub

        Protected Overrides Sub OnClosing(ByVal e As CancelEventArgs)
            isEnabled = False
            MyBase.OnClosing(e)
        End Sub

        Public Class DataPoint

            Public Sub New(ByVal argument As Date, ByVal value As Double)
                Me.Argument = argument
                Me.Value = value
            End Sub

            Public Property Argument As Date

            Public Property Value As Double
        End Class
    End Class
End Namespace
